src/core/lib/slice/slice_internal.h

```diff
@@ -23,17 +23,168 @@
 
 #include <grpc/slice.h>
 #include <grpc/slice_buffer.h>
+#include <string.h>
 
-inline const grpc_slice& grpc_slice_ref_internal(const grpc_slice& slice) {
+#include "src/core/lib/gpr/murmur_hash.h"
+#include "src/core/lib/gprpp/ref_counted.h"
+#include "src/core/lib/transport/static_metadata.h"
+
+extern uint32_t static_metadata_hash_values[GRPC_STATIC_MDSTR_COUNT];
+extern uint32_t g_hash_seed;
+namespace grpc_core {
+
+class SliceRefcount {
+ public:
+  enum Type { STATIC, INTERNED, REGULAR };
+  typedef void (*DestroyerFn)(void*);
+
+  SliceRefcount() = default;
+  SliceRefcount(SliceRefcount::Type type, RefCount* ref, bool ensureRefNonZero,
+                DestroyerFn destroyer_fn, void* destroyer_arg,
+                grpc_slice_refcount* sub)
+      : sub_refcount_(sub),
+        ref_type_(type),
+        ref_(ref),
+#ifndef NDEBUG
+        ensureRefNonZero_(ensureRefNonZero),
+#endif
+        dest_fn_(destroyer_fn),
+        destroy_fn_arg_(destroyer_arg) {
+    // Nothing to do here.
+  }
+  SliceRefcount(grpc_slice_refcount* sub, Type type)
+      : sub_refcount_(sub), ref_type_(type) {
+    // Initializer for static.
+  }
+
+  Type getType() const { return ref_type_; }
+
+  int Eq(const grpc_slice& a, const grpc_slice& b);
+
+  uint32_t Hash(const grpc_slice& slice);
+  void Ref() {
+    if (!ref_) return;
+#ifndef NDEBUG
+    if (ensureRefNonZero_) {
+      ref_->RefNonZero();
+      return;
+    }
+#endif
+    ref_->Ref();
+  }
+  void Unref() {
+    if (!ref_) return;
+    if (ref_->Unref()) {
+      dest_fn_(destroy_fn_arg_);
+    }
+  }
+
+  bool Validate() const { return ref_ != nullptr; }
+
+ private:
+  friend struct ::grpc_slice_refcount;
+
+  grpc_slice_refcount* sub_refcount_ = nullptr;
+  Type ref_type_ = REGULAR;
```

Comment:
this should be const, and probably don't need an initializer.

Reply:
Made it const.